### ISSUE TITLE
[GENERAL SUPPORT]: Updating trial parameters after get_next_trial

### ISSUE THREAD

**Issue by davifebba** [facebook/Ax#3061]
### Question

I have the following problem (using the **ServiceAPI**): I have time and temperature as search parameters. When I use get_next_trial(), AxClient suggests both, which is fine for temperature. However, I need to mark the trial as completed after the evaluation method, so that the suggested time is no longer valid and needs to be updated to the correct time after the evaluation method is called. For example:

```
next_parameters, trial_index = ax_client.get_next_trial()
next_parameters["time"] = 2.0 # for example
evaluation_method() # At this point, the time has passed. New time is new_time = 2.0 + delta_t
ax_client.complete_trial(...) # The updated time (new_time) must be informed to the client.
```

How to solve this problem? Using fixed_features in get_next_trial does not work because I need the updated time. So how to update the parameterization with an adjusted parameter before marking the trial as complete, using a generation strategy? In other words, how to update the arm parameters in the generation strategy before marking the trial as complete?

I'm using this generation strategy, so that Botorch takes over after a few sobol samples. If I use ax_client.attach_trial(), then it's a manual step and not part of the generation strategy.

```
GenerationStep(  # Initialization step
    # Which model to use for this step
    model=Models.SOBOL,
    # How many generator runs (each of which is then made a trial) to produce with this step
    num_trials=sobol_trials,
    # How many trials generated from this step must be COMPLETED before the next one
    min_trials_observed=sobol_trials,
),
GenerationStep(  # BayesOpt step
    model=Models.BOTORCH_MODULAR,
    # No limit on how many generator runs will be produced
    num_trials=botorch_trials,
    model_kwargs=self.botorch_model_kwargs,
),
])

```

### Code of Conduct

- [X] I agree to follow this Ax's Code of Conduct

**Comment by davifebba** [comment]
Hi @mgrange1998 !

2. I'm measuring the current-voltage response of a device, which degrades overtime. The idea is to use time and temperature as a parameter to build an interpolation model to predict device response as a function of these two parameters. The device not only depends on temperature, but also on its age. I'm trying to build a model that can make predictions as a function of (temperature, time) in an active learning scenario. If I use time as `fixed_features`, the model than can take temperature and time as independent variables and calculate the mean and uncertainty of the device response. When Ax suggests a new temperature with time as fixed feature, this would work just fine if I didn't have to wait for a temperature ramp to then measure the device response, at a different time than the one suggested.

1. Yes, it works with only temperature, and even with time using `ax.client.get_next_trial(fixed_features=FixedFeatures({'time':time.time()-start_time}))`
However, the time that goes into FixedFeatures has a -delta_T to the real time I wanted to pass to the model.

**Comment by Balandat** [comment]
Interesting. You're essentially trying to use time as a "contextual" variable. Unfortunately, I don't think this is easily supported in the current service API, but it's something that we want to support going forward. This would require quite a bit of additional work on our internals though, so having full support for this may take some time. 

That said, @sdaulton has done related things using a [`TimeAsFeature`](https://github.com/facebook/Ax/blob/63de39198c8b7aa0cad78bcf177c7ad64b9bebb7/ax/modelbridge/transforms/time_as_feature.py#L36-L46) transform. This is not the exactly what you'd need here since this uses the trial start and end times rather than the age of the device. What you'd need is some other transform that converts the "time" metric into an addiitonal parameter for the model.

**Comment by davifebba** [comment]
Exactly, time in this case would be a "contextual" variable. Since we cannot control time, the model would suggest only temperature points, but also use time to make predictions. In my case, "age" means the elapsed time since the start of the device characterization. Would it be possible to implement something using the Developer API, and informing the model of the correct time after a new trial has been suggested?

**Comment by Balandat** [comment]
> Would it be possible to implement something using the Developer API, and informing the model of the correct time after a new trial has been suggested?

Yes, that should be possible. We can implement a new `Transform` that (i) add the new time parameter as a parameter to the (transformed) search space and (ii) grabs the specified metric from the observation features and injects it as the parameter value instead. This will be very similar to the [`TimeAsFeature`](https://github.com/facebook/Ax/blob/63de39198c8b7aa0cad78bcf177c7ad64b9bebb7/ax/modelbridge/transforms/time_as_feature.py#L36-L46) transform. We would then have to specify this as a transform to use in a custom `GenerationStrategy` setup.

**Comment by davifebba** [comment]
@Balandat, any guidance or example on how to implement that would be very appreciated!

**Comment by sdaulton** [comment]
Hi @davifebba! It seems like using TimeAsFeature would work here. Although this uses the start time/duration as a contextual variables, the start time/duration should just be the device age with some additive offset, right?

For a given trial, you can report the start/end time for the evaluation to Ax by passing  `"start_time"` and `"end_time"` via the `metadata` dictionary in `AxClient.complete_trial` (https://github.com/facebook/Ax/blob/main/ax/service/ax_client.py#L770). If you then include the TimeAsFeature transform in the `BOTORCH_MODULAR` GenerationStep (e.g. by passing `model_kwargs={"transforms": Cont_X_trans+TimeAsFeature + Y_trans}`), then start time and duration will be added as contextual features to the model.

Does that address the problem or am I misunderstanding?

**Comment by davifebba** [comment]
Hi @sdaulton! Thanks for the help! So, what I want to accomplish in the end is to use temperature and time as independent variables to predict device performance. In this case, time is any real time instant in the future. Do you mean to implement something like this?

```
optimization_parameters = [
    {"name": "temperature", 
     "type": "range", 
     "bounds": [300, 550], 
     "value_type": "int"
     }
]
model_kwargs={"transforms": Cont_X_trans+[TimeAsFeature] + Y_trans}

BO loop:
   start_time = time.time()
   next_parameters, trial_index = self.optimizer.ax_client.get_next_trial()
   (time consuming steps)
   elapsed_time = (time.time()-start_time)/3600 # In hours
   complete_trial (...metadata={"start_time": start_time, "end_time": elapsed_time)
  ```
  
  I tried that and got this error, which I don't know what it means:
  `ax.exceptions.core.UserInputError: Upper bound of start_time must be strictly larger than lower.Got: (1.732642, 1.732642).`

**Comment by sdaulton** [comment]
Hi @davifebba, sorry for the delay. That looks good, but "end_time" should be the current time, not the elapsed time since the start.

**Comment by lena-kashtelyan** [comment]
Hi @davifebba, could you provide your full code, experiment setup and everything?

**Comment by davifebba** [comment]
Hi @lena-kashtelyan! Sure! Here's a complete example of what I'm trying to do. In this case I simulated the device's response as a function of time, temperature and gas concentration. The problem is that after calling `get_next_trial`, I need to run some time consuming steps, such as ramping up the heater to the suggested temperature, so by the time I call `complete_trial`, the time has changed. Thus, the model will be trained on incorrect time instants, since the actual device response measurement happens after the time consuming steps. In this case I was using `fixed_features`, but then I only get the time of suggestion, not completion. If I then try to use what @sdaulton suggested, I have the error I mentioned before:

`ax.exceptions.core.UserInputError: Upper bound of start_time must be strictly larger than lower.Got: (1.740674, 1.740674).`

Here's my full code:

`from botorch.acquisition.analytic import PosteriorStandardDeviation
from botorch.acquisition.input_constructors import acqf_input_constructor
from typing import Any, Dict
from botorch.models.model import Model
from botorch.acquisition.objective import PosteriorTransform
from ax.modelbridge.generation_strategy import GenerationStrategy, GenerationStep
from ax.modelbridge.registry import Models
from ax.service.ax_client import AxClient
from ax.service.ax_client import ObjectiveProperties
from ax.service.utils.instantiation import FixedFeatures
from ax.modelbridge.transforms.time_as_feature import TimeAsFeature
from ax.modelbridge.registry import Models, Cont_X_trans, Y_trans
from ax.models.torch.botorch_modular.surrogate import Surrogate
from botorch.models import SingleTaskGP
import time
import numpy as np


@acqf_input_constructor(PosteriorStandardDeviation)
def construct_inputs_PSD(model: Model, posterior_transform: PosteriorTransform | None = None, maximize: bool = True
                            ) -> Dict[str, Any]:
    return {"model": model}


def bayes_opt(optimization_parameters, evaluate, bo_options, std_noise=0):
    try:
        gs = GenerationStrategy(
            steps=[
                GenerationStep(model=Models.SOBOL, num_trials=bo_options['sobol_runs'], should_deduplicate=bo_options['sobol_deduplicate']),
                GenerationStep(
                    model=Models.BOTORCH_MODULAR, 
                    num_trials=-1,           
                    model_kwargs={"surrogate": bo_options['surrogate_model'],
                                "botorch_acqf_class": bo_options['acquisition_function'],
                                "transforms": Cont_X_trans+[TimeAsFeature] + Y_trans},
                    # model_kwargs={"surrogate": bo_options['surrogate_model'],
                    #             "botorch_acqf_class": bo_options['acquisition_function']},                                    
                                
                    should_deduplicate=bo_options['botorch_deduplicate']),
            ]
        )
        print("Generation strategy created successfully")
    except:
        print(f"Failed to create generation strategy:")

    ax_client = AxClient(generation_strategy=gs, verbose_logging=False)
    ax_client.create_experiment(
        name="Active Learning", 
        parameters=optimization_parameters, 
        objectives={"Objective": ObjectiveProperties(minimize=True)}
    )

    start_time = time.time()
    for i in range(bo_options['max_runs']):
        # parameters, trial_index = ax_client.get_next_trial(fixed_features=FixedFeatures({'x1':time.time()-start_time}))
        parameters, trial_index = ax_client.get_next_trial(fixed_features=FixedFeatures({'x1':time.time()-start_time}))
        print(f"Iteration {i}: {parameters}")
        time.sleep(2) # Simulate a time consuming step, for example ramping up the temperature
        end_time = time.time()
        raw_data = evaluate(parameters, std_noise=std_noise) 
        # Ideally update the trial parameters before completing the trial.
        ax_client.complete_trial(trial_index=trial_index, raw_data=raw_data, metadata={"start_time": start_time, "end_time": end_time})

    # Refit the model before plotting.
    ax_client.generation_strategy._fit_current_model(data=ax_client.experiment.lookup_data())
    model = ax_client.generation_strategy.model
    print("BO done!")
    return model

def simulate_equilibrium_response(parameters, std_noise=0):
    """
    Simulates a noisy equilibrium sensor response with temperature-dependent sensitivity.

    Parameters:
        T : Temperature (°C)
        C : Gas concentration (ppm)
        t : Time (minutes)
        std_noise : Percentage of the measured value used as noise standard deviation (default 0)

    Returns:
        Noisy sensor response (arbitrary units)
    """
    t = parameters["x1"]
    T = parameters["x2"]
    C = parameters["x3"]        
    # Constants
    R = 8.314  # Gas constant in J/(mol*K)
    Ea = 5000  # Activation energy in J/mol

    # Temperature effect: Arrhenius law
    temp_effect = 1000 * np.exp(-Ea / (R * (T + 273.15)))

    # --- Enhanced Gas Concentration Effect with Temperature Dependence ---
    gas_effect = (10 * (C ** 0.5) * (1 - np.exp(-C / 10))) * (1 + 0.01 * (T - 300))

    # --- Time Effect ---
    alpha = 0.01  # Decay rate for the exponential part
    weight_constant = 0.7
    weight_decay = 0.3
    time_factor = weight_constant + weight_decay * np.exp(-alpha * t)

    # Linear drift (subtractive)
    drift = 0.001 * t

    # Daily periodic variation
    periodic = 0.5 * np.sin(2 * np.pi * t / 1440)

    # --- Combine Effects ---
    raw_response = (temp_effect + gas_effect) * time_factor - drift + periodic

    # Add noise as a percentage of the measured value
    noise_std = std_noise * np.abs(raw_response)  # Standard deviation is a percentage of the response
    noise = np.random.normal(0, noise_std, size=np.shape(raw_response))

    noisy_response = raw_response + noise

    return noisy_response      

# Define optimization parameters for the 1D BO experiment.
optimization_parameters = [
    {
        "name": "x1", # time
        "type": "range",
        "bounds": [0, 500],
        "value_type": "float",
        "log_scale": False,
    },
    {
        "name": "x2", # temperature
        "type": "range",
        "bounds": [300, 600],
        "value_type": "int",
        "log_scale": False,
    },
    {
        "name": "x3", # concentration (ppm)
        "type": "range",
        "bounds": [0, 1500],
        "value_type": "int",
        "log_scale": False,
    }        

]

bo_options = {"surrogate_model": Surrogate(SingleTaskGP),
            "acquisition_function": PosteriorStandardDeviation,
            "sobol_runs": 2,"max_runs": 4,
            "sobol_deduplicate": False, "botorch_deduplicate": False}

bayes_opt(optimization_parameters, simulate_equilibrium_response, bo_options)
     `

**Comment by davifebba** [comment]
No worries. This is not an issue anymore. Thanks!
